File: windows/src/GoalMaker.Core/Planning/HabitHeat.cs
namespace GoalMaker.Core.Planning;

/// <summary>
/// One day of a habit's heatmap (docs/habits.md): <see cref="Kind"/> says whether there is nothing,
/// a pause, a skip, or a <see cref="Fraction"/> of the day's target from 0 to 1.
/// </summary>
public readonly record struct HabitHeat(HabitHeatKind Kind, double Fraction = 0)
{
    public static HabitHeat None => new(HabitHeatKind.None);

    public static HabitHeat Paused => new(HabitHeatKind.Paused);

    public static HabitHeat Skipped => new(HabitHeatKind.Skipped);

    public static HabitHeat Share(double fraction) => new(HabitHeatKind.Share, fraction);
}

File: android/app/src/main/kotlin/com/goalmaker/app/application/planning/HabitHeat.kt
package com.goalmaker.app.application.planning

/** One day of a habit's heatmap (docs/habits.md). */
sealed interface HabitHeat {
    /** Before the habit starts, or a day it isn't due. */
    data object None : HabitHeat

    data object Paused : HabitHeat

    data object Skipped : HabitHeat

    /** A day that went over a limit habit's number. */
    data object Over : HabitHeat

    /** The day's value against its target, 0 to 1. */
    data class Share(val fraction: Double) : HabitHeat
}
